File: src/test/java/org/swordess/persistence/json/java/JsonEntityMetadataTest.java
package org.swordess.persistence.json.java;

import kotlin.jvm.JvmClassMappingKt;
import org.junit.Test;
import org.swordess.persistence.Id;
import org.swordess.persistence.json.JsonEntityMetadata;
import org.swordess.persistence.json.java.test.brokenmodel.Foo;
import org.swordess.persistence.json.java.test.brokenmodel.Group;
import org.swordess.persistence.json.java.test.brokenmodel.User;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

public class JsonEntityMetadataTest {

    @Test
    public void testExceptionShouldBeThrownIfIdAnnotationAbsent() {
        try {
            new JsonEntityMetadata(JvmClassMappingKt.getKotlinClass(Foo.class));
            fail("exception expected");
        } catch (RuntimeException e) {
            assertEquals("Annotation of type " + Id.class.getName() + " should present at least once in " + Foo.class.getName(), e.getMessage());
        }
    }

    @Test
    public void testExceptionShouldBeThrownIfSetterForIdIsMissing() {
        try {
            new JsonEntityMetadata(JvmClassMappingKt.getKotlinClass(User.class));
            fail("exception expected");
        } catch (RuntimeException e) {
            assertEquals("no setter was found for property id, please add a setter", e.getMessage());
        }
    }

    @Test
    public void testExceptionShouldBeThrownIfGetterForIdIsNeitherPrefixedWithGetNorIs() {
        try {
            new JsonEntityMetadata(JvmClassMappingKt.getKotlinClass(Group.class));
            fail("exception expected");
        } catch (RuntimeException e) {
            assertEquals(
                    "unable to determine property name for public java.lang.Long org.swordess.persistence.json.java.test.brokenmodel.Group.fetchId()",
                    e.getMessage()
            );
        }
    }

}

File: src/test/kotlin/org/swordess/persistence/json/JsonEntityMetadataTest.kt
package org.swordess.persistence.json

import org.junit.Test
import org.swordess.persistence.Id
import org.swordess.persistence.json.java.test.brokenmodel.User
import org.swordess.persistence.json.test.brokenmodel.Foo
import kotlin.test.assertEquals
import kotlin.test.fail

class JsonEntityMetadataTest {

    @Test
    fun testExceptionShouldBeThrownIfIdAnnotationAbsent() {
        try {
            JsonEntityMetadata(Foo::class)
            fail("exception expected")
        } catch (e: RuntimeException) {
            assertEquals("Annotation of type ${Id::class.java.name} should present at least once in ${Foo::class.java.name}", e.message)
        }
    }

    @Test
    fun testExceptionShouldBeThrownIfSetterForIdIsMissing() {
        try {
            JsonEntityMetadata(User::class)
            fail("exception expected")
        } catch (e: RuntimeException) {
            assertEquals("no setter was found for property id, please add a setter", e.message)
        }
    }

}